lib/galaxy/tools/actions/__init__.py

```diff
@@ -748,17 +751,24 @@ class ObjectStorePopulator(object):
     def __init__(self, app):
         self.object_store = app.object_store
         self.object_store_id = None
+        self.objects = []
 
-    def set_object_store_id(self, data):
+    def register(self, data):
         # Create an empty file immediately.  The first dataset will be
         # created in the "default" store, all others will be created in
         # the same store as the first.
-        data.dataset.object_store_id = self.object_store_id
+        dataset = data.dataset
+        dataset.object_store_id = self.object_store_id
+        self.objects.append(dataset)
         try:
-            self.object_store.create(data.dataset)
+            self.object_store.set_object_store_id(dataset)
         except ObjectInvalid:
-            raise Exception('Unable to create output dataset: object store is full')
-        self.object_store_id = data.dataset.object_store_id  # these will be the same thing after the first output
+            raise Exception('Unable to set object store id for output dataset: object store is full')
+        self.object_store_id = dataset.object_store_id  # these will be the same thing after the first output
```

Comment:
I think lines 761-767 could be simplified to:
```python
if self.objest_store_id:
    dataset.object_store_id = self.object_store_id
else:
    self.object_store.set_object_store_id(dataset)
    self.object_store_id = dataset.object_store_id  # these will be the same thing after the first output
```

See also comment below.

Reply:
I can see why this is less confusing. That said I setup the abstraction on purpose. Letting the object store itself always see the dataset even when the ``object_store_id`` is set is useful if we ever do additional things with that abstraction such as add additional fields based on the primary ID (credentials, etc..) or if the object store does counting of objects or something to determine when it is "full". So it feels like a good abstraction to me - but I can see thinking it is speculative and convoluted. Also the cleanup isn't as nice if you account https://github.com/galaxyproject/galaxy/pull/6373#discussion_r204827609 and that the exception handling still needs to be here I think.

So I think I like this on the whole better the way it is now, but if you feel strong I can make the change - I think it would look like...

```python
    def register(self, data):
        # Create an empty file immediately.  The first dataset will be
        # created in the "default" store, all others will be created in
        # the same store as the first.
        dataset = data.dataset
        if self.object_store is not None:
            dataset.object_store_id = self.object_store_id
        else:
            try:
                self.object_store.set_object_store_id(dataset)
            except ObjectInvalid:
                raise Exception('Unable to set object store id for output dataset: object store is full')
            self.object_store_id = dataset.object_store_id  # these will be the same thing after the first output

        self.objects.append(dataset)
```

Reply:
1) There is at least another place in the codebase where the `object_store_id` is assigned directly: lib/galaxy/tools/actions/upload_common.py
2) I think that the proper abstraction for what you want to do then would require a second optional parameter for the `set_object_store_id()` method to specify a value (default `None`).